ops/src/migrate_slack.rs

```diff
@@ -0,0 +1,61 @@
+use crate::slack::Slack;
+use octobot_lib::config::Config;
+use octobot_lib::errors::*;
+
+pub async fn migrate_slack_id(config: &Config, slack: &Slack) {
+    match do_migrate_slack_id(config, slack).await {
+        Ok(count) => {
+            if count > 0 {
+                log::info!("Migrated {} slack users", count);
+            }
+        }
+        Err(e) => {
+            log::error!("Failed to migrate slack users: {}", e);
+        }
+    }
+}
+
+async fn do_migrate_slack_id(config: &Config, slack: &Slack) -> Result<u32> {
+    let mut all_users = config
+        .users()
+        .get_all()?
+        .into_iter()
+        .filter(|u| u.slack_id.is_empty())
+        .collect::<Vec<_>>();
+
+    if all_users.is_empty() {
+        return Ok(0);
+    }
+
+    let all_slack_users = slack.list_users().await?;
+
+    let mut migrated = 0;
+    for user in &mut all_users {
+        for slack in &all_slack_users {
+            if user.slack_name == slack.name {
+                user.slack_id = slack.id.clone();
+                user.email = slack.profile.email.clone();
+
+                if let Err(e) = config.users_write().update(user) {
+                    log::error!(
+                        "Failed to migrate slack name: {} -> {}: {}",
+                        slack.name,
+                        slack.id,
+                        e
+                    );
+                } else {
+                    log::info!("Migrated slack name: {} -> {}", slack.name, slack.id);
+                    migrated += 1;
+                }
+            }
+        }
+    }
+
+    for user in &all_users {
+        if user.slack_id.is_empty() {
+            log::info!("No matched slack username: {}", user.slack_name)
```

Comment:
This isn't treated as an error, and AFAICT the migration runs every time we find a user without a username. So if for some reason there is a user without a username, we'll go through this migration again every time we start up.

That's probably not a big deal, but I figured I'd call it out. I'm not sure if it's worth adding a migration flag to say "we've already tried this"

Reply:
yeah, indeed. i was trying to think of a way to avoid doing this every time. could write something to the config.toml maybe?

Reply:
It feels like we shouldn't mutate the user-written config.toml, could it just be written to the config DB instead? Though I suppose a config.toml flag would allow you to unset it to manually force a retry.

Alternatively, I guess we could check if _any_ user has an ID, which would imply we'd migrated, though we'd probably want to wrap the above in a transaction, so we'd only set ID's if we had a full successful run.

Reply:
i was gonna solve this actually by just removing/editing those users on first migrate that failed.

...then remove this code later on. 😬 

thoughts?